beluga_amcl/test/test_amcl_node.cpp

```diff
@@ -0,0 +1,861 @@
+// Copyright 2023 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <tf2_ros/create_timer_ros.h>
+
+#include <beluga_amcl/private/amcl_node.hpp>
+
+#include <lifecycle_msgs/msg/state.hpp>
+
+namespace
+{
+
+using namespace std::chrono_literals;
+
+/// Spin a group of nodes until a condition is met.
+/**
+ * \param predicate The stop condition.
+ * \param timeout Maximum time to spin.
+ * \param nodes The nodes to spin.
+ * \return True if the condition was met. False if it timed out.
+ */
+template<class Predicate, class Rep, class Period, class ... Nodes>
+bool spin_until(
+  Predicate && predicate,
+  const std::chrono::duration<Rep, Period> & timeout,
+  const std::shared_ptr<Nodes> & ... nodes)
+{
+  rclcpp::executors::SingleThreadedExecutor executor;
+  (executor.add_node(nodes->get_node_base_interface()), ...);
+
+  auto start = std::chrono::high_resolution_clock::now();
+  while (std::chrono::high_resolution_clock::now() - start < timeout) {
+    executor.spin_some();
+    if (predicate()) {
+      return true;
+    }
+  }
+  return false;  // timed out
```

Comment:
@nahueespinosa this can busy loop if there's nothing in queue. Consider replicating [`rclcpp::Executor::spin_until_future_complete`](https://github.com/ros2/rclcpp/blob/f8072f2fa265000a2a1be8f3dbe7e04d9dfb1c91/rclcpp/include/rclcpp/executor.hpp#L345) implementation e.g.:

```suggestion
  auto deadline = std::chrono::high_resolution_clock::now() + timeout;
  while (rclcpp::ok() && !predicate() && std::chrono::high_resolution_clock::now() < deadline) {
      executor.spin_once(deadline - std::chrono::high_resolution_clock::now());  // wait for it
      executor.spin_some(deadline - std::chrono::high_resolution_clock::now());  // flush it all out
  }
  return predicate();  // last minute check
```

Reply:
Done in https://github.com/Ekumen-OS/beluga/pull/194/commits/00b9631edb630940b1e87c2e57e0919da9320a67.